File: frontend/src/api/route.ts
import { API_BASE_URL } from './reports'
import type { Report } from './reports'

export interface AddressPoint {
  lat: number
  lng: number
  address?: string
}

export interface HazardPoint {
  id: string
  latitude: number
  longitude: number
  risk_score: number
  risk_factors: string[]
  accident_count: number
  osm_tags: Record<string, string>
  title?: string | null
  distance_from_origin_m?: number | null
}

export interface RouteDangerReport {
  report: Report
  distance_from_route_m: number
}

export interface RouteResponse {
  route_geometry: {
    type: 'LineString'
    coordinates: [number, number][]
  }
  distance_m: number
  duration_s: number
  hazard_points: HazardPoint[]
  danger_reports: RouteDangerReport[]
}

export interface GeocodeResult {
  name: string
  lat: number
  lng: number
}

async function parseErrorDetail(response: Response, fallback: string): Promise<string> {
  const body = await response.json().catch(() => null)
  return (body && typeof body.detail === 'string' && body.detail) || fallback
}

export async function geocode(query: string, limit = 5): Promise<GeocodeResult[]> {
  const trimmed = query.trim()
  if (!trimmed) return []

  const params = new URLSearchParams({ q: trimmed, limit: String(limit) })
  const response = await fetch(`${API_BASE_URL}/api/geocode?${params.toString()}`)

  if (!response.ok) {
    throw new Error(await parseErrorDetail(response, '地点検索に失敗しました'))
  }

  return response.json()
}

export async function createRoute(origin: AddressPoint, destination: AddressPoint): Promise<RouteResponse> {
  const response = await fetch(`${API_BASE_URL}/api/route`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ origin, destination }),
  })

  if (!response.ok) {
    throw new Error(await parseErrorDetail(response, 'ルート検索に失敗しました'))
  }

  return response.json()
}

File: backend/app/models/route.py
from typing import Literal

from pydantic import BaseModel

from app.models.report import Report


class GeocodeResult(BaseModel):
    name: str
    latitude: float
    longitude: float


class RoutePoint(BaseModel):
    latitude: float
    longitude: float


class RouteResult(BaseModel):
    points: list[RoutePoint]
    distance_m: float
    duration_s: float


class RouteDangerReport(BaseModel):
    report: Report
    distance_from_route_m: float


class RouteLocation(BaseModel):
    lat: float
    lng: float
    address: str | None = None


class RouteRequest(BaseModel):
    origin: RouteLocation
    destination: RouteLocation


class HazardPoint(BaseModel):
    id: str
    latitude: float
    longitude: float
    risk_score: int
    risk_factors: list[str]
    accident_count: int
    osm_tags: dict[str, str]
    title: str | None = None
    distance_from_origin_m: float | None = None


class RouteGeometry(BaseModel):
    type: str = "LineString"
    coordinates: list[list[float]]


class RouteOption(BaseModel):
    id: str
    kind: Literal["recommended", "alternative"]
    route_geometry: RouteGeometry
    distance_m: float
    duration_s: float
    safety_score: int
    safety_grade: Literal["A", "B", "C", "D", "E"]
    hazard_points: list[HazardPoint]
    danger_reports: list[RouteDangerReport]


class RouteResponse(BaseModel):
    routes: list[RouteOption]


class GeocodeItem(BaseModel):
    name: str
    lat: float
    lng: float
